# **Importing Libraries and Reading Data**

In [54]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [55]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
from tabulate import tabulate

In [56]:
df_customers = pd.read_csv("/content/drive/MyDrive/3rd year 2024-25/Term 2/Machine Learning and Predictive Analytics/raw data/customers_data.csv")
df_products = pd.read_csv("/content/drive/MyDrive/3rd year 2024-25/Term 2/Machine Learning and Predictive Analytics/raw data/products_data.csv")
df_transactions = pd.read_csv("/content/drive/MyDrive/3rd year 2024-25/Term 2/Machine Learning and Predictive Analytics/raw data/transactions_data.csv")

# **Data Overview**

In [57]:
# Dictionary of DataFrames
dfs = {
    "customers": df_customers,
    "products": df_products,
    "transactions": df_transactions
}

### Data Types

In [58]:
for name, df in dfs.items():
    print(f"Data types for {name}:")
    print(df.dtypes)
    print()

Data types for customers:
Company_ID        float64
Company_Name       object
Company_Profit    float64
Address            object
dtype: object

Data types for products:
Product_ID       float64
Product_Name      object
Product_Price     object
dtype: object

Data types for transactions:
Unnamed: 0          float64
Transaction_ID      float64
Company_ID          float64
Product_ID          float64
Quantity            float64
Transaction_Date     object
Product_Price       float64
Total_Cost          float64
dtype: object



### Summary Statistics

In [59]:
for name, df in dfs.items():
    print(f"{name}:\n{df.describe()}\n")

customers:
       Company_ID  Company_Profit
count   90.000000       88.000000
mean    51.944444    76400.500000
std     28.684397    27296.169253
min      1.000000    30663.000000
25%     27.500000    52648.500000
50%     53.500000    75301.500000
75%     76.750000   101089.750000
max    100.000000   118114.000000

products:
       Product_ID
count   18.000000
mean    10.333333
std      6.077925
min      1.000000
25%      5.250000
50%     10.500000
75%     14.750000
max     20.000000

transactions:
        Unnamed: 0  Transaction_ID   Company_ID   Product_ID     Quantity  \
count  9000.000000     9000.000000  9000.000000  9000.000000  9000.000000   
mean   4994.049111     5005.177556    50.583556    10.446778    10.528778   
std    2885.331476     2894.971964    28.900870     5.768341     5.808062   
min       0.000000        1.000000     1.000000     1.000000     0.000000   
25%    2501.750000     2491.750000    25.000000     5.000000     6.000000   
50%    4997.500000     5005.00000

### Number of rows and columns in the DataFrame

In [60]:
for name, df in dfs.items():
    print(f"{name}: {df.shape}")

customers: (100, 4)
products: (20, 3)
transactions: (10000, 8)


### First 5 rows of each data set


In [61]:
for name, df in dfs.items():
    print(f"{name}:\n{tabulate(df.head(), headers='keys', tablefmt='psql')}\n")

customers:
+----+--------------+----------------------+------------------+------------------------------------------------------+
|    |   Company_ID | Company_Name         |   Company_Profit | Address                                              |
|----+--------------+----------------------+------------------+------------------------------------------------------|
|  0 |            1 | Tech  Enterprises  1 |            80701 | EDSA, Barangay 606, Pasig, Philippines               |
|  1 |            2 | Global  Partners  2  |            80511 | Commonwealth Ave, Barangay 789, Taguig, Philippines  |
|  2 |            3 | Quantum Associates 3 |           110664 | Roxas Blvd, Barangay 505, Pasig, Philippines         |
|  3 |            4 | Prime Network 4      |              nan | Alabang-Zapote Rd, Barangay 202, Taguig, Philippines |
|  4 |            5 | Elite  Ventures  5   |            69427 | Ayala Avenue, Barangay 101, Makati, Philippines      |
+----+--------------+----------------

### Last 5 rows of each data set

In [62]:
for name, df in dfs.items():
    print(f"{name}:\n{tabulate(df.tail(), headers='keys', tablefmt='psql')}\n")

customers:
+----+--------------+----------------------+------------------+------------------------------------------------------+
|    |   Company_ID | Company_Name         |   Company_Profit | Address                                              |
|----+--------------+----------------------+------------------+------------------------------------------------------|
| 95 |           96 | Dynamic  Network  96 |           101428 | Alabang-Zapote Rd, Brgy. 456, Cebu City, Philippines |
| 96 |           97 | Quantum Holdings 97  |            33449 | EDSA, Barangay 789, Manila, Philippines              |
| 97 |           98 | Pioneer Ventures 98  |            71095 | Roxas Blvd, Barangay 123, Taguig, Philippines        |
| 98 |          nan | Elite Corp 99        |           107929 | Alabang-Zapote Rd, Barangay 303, Makati, Philippines |
| 99 |          100 | Global Solutions 100 |              nan | Ayala avenue, barangay 456, taguig, philippines      |
+----+--------------+----------------

###Display Index, Columns, and Data

In [63]:
for name, df in dfs.items():
  print(f"DataFrame Info for {name}:")
  df.info()
  print()

DataFrame Info for customers:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Company_ID      90 non-null     float64
 1   Company_Name    100 non-null    object 
 2   Company_Profit  88 non-null     float64
 3   Address         100 non-null    object 
dtypes: float64(2), object(2)
memory usage: 3.3+ KB

DataFrame Info for products:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Product_ID     18 non-null     float64
 1   Product_Name   20 non-null     object 
 2   Product_Price  20 non-null     object 
dtypes: float64(1), object(2)
memory usage: 612.0+ bytes

DataFrame Info for transactions:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 colum

**`customers_data.csv`** - This dataset contains information about the companies, including their names, profits, and addresses.

|Column Name|Description|Data Type|Sample Data|
|:---|:---|:---|:---|
|Company_ID|Unique identifier for the company|float|1|
|Company_Name|Name of the company|obj|Tech Enterprises 1|
|Company_Profit|Profit of the company|float|80701|
|Address|Address of the company|obj|EDSA, Barangay 606, Pasig, Philippines|

Observations:


*   Company_ID: Missing a few IDs, need to be filled in order. Chaange float to int.
*   Company_Name: Need to be cleaned, Company_ID is included in the end.
*   Company_Profit: No currency indicated. Fill in missing data or drop or keep it blank?
*   Address: Need to standardize format (Brngy vs Barangay). Fix Capitalization.




**`products_data.csv`** - This dataset contains information about the products, including their names and prices.

|Column Name|Description|Data Type|Sample Data|
|:---|:---|:---|:---|
|Product_ID|Unique identifier for the product|float|1|
|Product_Name|Name of the product|obj|FinPredictor Suite|
|Product_Price|Price of the product|obj|?140,000|

Observations:


*   Product_ID: Missing a few IDs, need to be filled in order.
*   Product_Name: No issues.
*   Product_Price: What currency? Need to standardize format.

**`transactions_data.csv`** - This dataset contains information about the transactions, including the company and product IDs, quantities, prices, and dates.

|Column Name|Description|Data Type|Sample Data|
|:---|:---|:---|:---|
|Unnamed: 0|Index of the row|float|0|
|Transaction_ID|Unique identifier for the transaction|float|1|
|Company_ID|Unique identifier for the company|float|88|
|Product_ID|Unique identifier for the product|float|6|
|Quantity|Quantity of the product|float|16|
|Transaction_Date|Date of the transaction|obj|2024/03/26|
|Product_Price|Price of the product|float|194379|
|Total_Cost|Total cost of the transaction|float|1.0752e+06|

Observations:
*   Unnamed: 0: Seems to be index of rows since  t starts with 0, can be dropped.
*   Transaction_ID: Missing data, seems to follow a sequential order (1,2,3,4, etc.). Change data type to int.
*   Company_ID: Missing Data. What to do? Drop data? Cannot be filled in since this is linked to the Company who purchased.
*   Product_ID: Matches the `Product_ID` column in the `products_data.csv` file.


Link with previous data:
*   The `Company_ID` column in the `transactions_data.csv` file matches the `Company_ID` column in the `customers_data.csv` file.

*   The `Product_ID` column in the `transactions_data.csv` file matches the `Product_ID` column in the `products_data.csv` file.
    """
    )

# **Data Cleaning**

In [64]:
customers = df_customers.copy()
products = df_products.copy()
transactions = df_transactions.copy()

## customers_data - data cleaning



In [65]:
# Fill missing Company_ID sequentially using the length of the customers DataFrame
customers['Company_ID'] = range(1, len(customers) + 1)

print(customers.isnull().sum())

Company_ID         0
Company_Name       0
Company_Profit    12
Address            0
dtype: int64


In [66]:
# Convert `Address` to title case
customers["Address"] = customers["Address"].str.title()

# Replace 'Barangay' with 'Brgy.'
customers["Address"] = customers["Address"].str.replace(r"\bBarangay\b", "Brgy.", regex=True)

# Replace 'Avenue' with 'Ave.'
customers["Address"] = customers["Address"].str.replace(r"\bAvenue\b", "Ave.", regex=True)
customers["Address"] = customers["Address"].str.replace(r"\bAve\b(?!\.)", "Ave.", regex=True)

# Replace 'Blvd' with 'Blvd.'
customers["Address"] = customers["Address"].str.replace(r"\bBlvd\b(?!\.)", "Blvd.", regex=True)

# Replace 'Rd' with 'Rd.'
customers["Address"] = customers["Address"].str.replace(r"\bRd\b(?!\.)", "Rd.", regex=True)

# Replace 'Slex' with 'SLEX'
customers["Address"] = customers["Address"].str.replace(r"\bSlex\b", "SLEX", regex=True)

# Replace 'Edsa' with 'EDSA'
customers["Address"] = customers["Address"].str.replace(r"\bEdsa\b", "EDSA", regex=True)

# Remove '!'
customers["Address"] = customers["Address"].str.replace("!", "", regex=True)

customers


,Company_ID,Company_Name,Company_Profit,Address
0,1,Tech Enterprises 1,80701.0,"EDSA, Brgy. 606, Pasig, Philippines"
1,2,Global Partners 2,80511.0,"Commonwealth Ave., Brgy. 789, Taguig, Philippines"
2,3,Quantum Associates 3,110664.0,"Roxas Blvd., Brgy. 505, Pasig, Philippines"
3,4,Prime Network 4,NaN,"Alabang-Zapote Rd., Brgy. 202, Taguig, Philipp..."
4,5,Elite Ventures 5,69427.0,"Ayala Ave., Brgy. 101, Makati, Philippines"
...,...,...,...,...
95,96,Dynamic Network 96,101428.0,"Alabang-Zapote Rd., Brgy. 456, Cebu City, Phil..."
96,97,Quantum Holdings 97,33449.0,"EDSA, Brgy. 789, Manila, Philippines"
97,98,Pioneer Ventures 98,71095.0,"Roxas Blvd., Brgy. 123, Taguig, Philippines"
98,99,Elite Corp 99,107929.0,"Alabang-Zapote Rd., Brgy. 303, Makati, Philipp..."


In [67]:
# Remove the digits from the `Company_Name` column
customers["Company_Name"] = customers["Company_Name"].str.replace("\d+", "", regex=True)

customers

,Company_ID,Company_Name,Company_Profit,Address
0,1,Tech Enterprises,80701.0,"EDSA, Brgy. 606, Pasig, Philippines"
1,2,Global Partners,80511.0,"Commonwealth Ave., Brgy. 789, Taguig, Philippines"
2,3,Quantum Associates,110664.0,"Roxas Blvd., Brgy. 505, Pasig, Philippines"
3,4,Prime Network,NaN,"Alabang-Zapote Rd., Brgy. 202, Taguig, Philipp..."
4,5,Elite Ventures,69427.0,"Ayala Ave., Brgy. 101, Makati, Philippines"
...,...,...,...,...
95,96,Dynamic Network,101428.0,"Alabang-Zapote Rd., Brgy. 456, Cebu City, Phil..."
96,97,Quantum Holdings,33449.0,"EDSA, Brgy. 789, Manila, Philippines"
97,98,Pioneer Ventures,71095.0,"Roxas Blvd., Brgy. 123, Taguig, Philippines"
98,99,Elite Corp,107929.0,"Alabang-Zapote Rd., Brgy. 303, Makati, Philipp..."


In [68]:
# Check for duplicates in the 'Company_Name' column
duplicates = customers[customers.duplicated(subset=['Company_Name'], keep=False)]

# Print the number of duplicate company names
num_duplicates = len(duplicates)
print(f"Number of duplicate company names: {num_duplicates}")

# Print the duplicate company names
if num_duplicates > 0:
    print("\nDuplicate company names:")
    print(duplicates['Company_Name'].unique())
else:
    print("\nNo duplicate company names found.")

Number of duplicate company names: 47

Duplicate company names:
['Tech  Enterprises  ' 'Quantum Associates ' 'Prime Network '
 'Dynamic Solutions ' 'Pioneer Network ' 'Dynamic Ventures '
 'Quantum Network ' 'Sky Network ' 'Sky Ent.s ' 'Tech  Solutions  '
 'Tech Associates ' 'Elite Corp ' 'Quantum Holdings ' 'Sky Consulting '
 'Prime  Consulting  ' 'Pioneer Corp ' 'Prime Holdings '
 'Innovative Holdings ' 'Prime Industries ' 'Tech Ventures '
 'Global Solutions ' 'Quantum Solutions ']


In [69]:
# Find rows containing "Quantum Network" in any column
rows_with_quantum_network = df_customers[customers.apply(lambda row: row.astype(str).str.contains('Quantum Network').any(), axis=1)]

# Print the rows with column names
print(rows_with_quantum_network.to_string(index=False))

 Company_ID       Company_Name  Company_Profit                                               Address
       12.0 Quantum Network 12         91671.0 Katipunan Ave, Barangay 123, Mandaluyong, Philippines
       43.0 Quantum Network 43        113336.0     Ortigas Ave, Barangay 404, Cebu City, Philippines


## products_data - data cleaning


In [70]:
products

,Product_ID,Product_Name,Product_Price
0,1.0,FinPredictor Suite,"?140,000"
1,2.0,MarketMinder Analytics,"?168,000"
2,3.0,TrendWise Forecaster,"?100,800"
3,4.0,CustomerScope Insights,"?123,200"
4,5.0,SalesSync Optimizer,"?84,000"
5,6.0,RevenueVue Dashboard,"?179,200"
6,7.0,DataBridge Integration Tool,"?151,200"
7,NaN,RiskRadar Monitor,"?151,200"
8,9.0,Product 9,"?112,000"
9,10.0,SegmentX Targeting,"?89,600"


In [71]:
# Fill missing Product_ID sequentially
products['Product_ID'] = range(1, len(products) + 1)

# Remove question mark from Product_Price
products['Product_Price'] = products['Product_Price'].str.replace('?', '', regex=False)

# Print updated DataFrame to verify
products

,Product_ID,Product_Name,Product_Price
0,1,FinPredictor Suite,"140,000"
1,2,MarketMinder Analytics,"168,000"
2,3,TrendWise Forecaster,"100,800"
3,4,CustomerScope Insights,"123,200"
4,5,SalesSync Optimizer,"84,000"
5,6,RevenueVue Dashboard,"179,200"
6,7,DataBridge Integration Tool,"151,200"
7,8,RiskRadar Monitor,"151,200"
8,9,Product 9,"112,000"
9,10,SegmentX Targeting,"89,600"
